iroh/src/ticket/blob.rs

```diff
@@ -0,0 +1,151 @@
+//! Tickets for blobs.
+
+use anyhow::{ensure, Result};
+use iroh_bytes::{protocol::RequestToken, BlobFormat, Hash};
+use iroh_net::{derp::DerpMap, key::SecretKey, PeerAddr};
+use serde::{Deserialize, Serialize};
+
+use crate::dial::Options;
+
+use super::*;
+
+/// A token containing everything to get a file from the provider.
+///
+/// It is a single item which can be easily serialized and deserialized.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, derive_more::Display)]
+#[display("{}", IrohTicket::serialize(self))]
+pub struct Ticket {
+    /// The provider to get a file from.
+    node: PeerAddr,
+    /// The format of the blob.
+    format: BlobFormat,
+    /// The hash to retrieve.
+    hash: Hash,
+    /// Optional Request token.
+    token: Option<RequestToken>,
+}
+
+impl IrohTicket for Ticket {
+    const KIND: Kind = Kind::Blob;
+}
+
+impl std::str::FromStr for Ticket {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        IrohTicket::deserialize(s)
+    }
+}
+
+impl Ticket {
+    /// Creates a new ticket.
+    pub fn new(
+        peer: PeerAddr,
+        hash: Hash,
+        format: BlobFormat,
+        token: Option<RequestToken>,
+    ) -> Result<Self> {
+        ensure!(
+            !peer.info.direct_addresses.is_empty(),
+            "addrs list can not be empty"
+        );
+        Ok(Self {
+            hash,
+            format,
+            node: peer,
+            token,
+        })
+    }
+
+    /// Deserializes from bytes.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self> {
```

Comment:
This shadowing of method names is a bit tricky. Maybe add a `verify_ticket` method to the trait (which is called when doing `from_bytes`, with a default impl of `Ok(())` and implement the verification you are doing here with that?